### PR TITLE
[dashboard] Disallow team names that might conflict with dashboard URLs

### PR DESCRIPTION
**Pull request by jankeromnes**
Fixes https://github.com/gitpod-io/gitpod/issues/4469

### How to test

1. Try to create a team called:
    - "workspaces" (or any other dashboard page)
    - "root" (or any other [suspicious name](https://github.com/marteinn/The-Big-Username-Blocklist/blob/main/list_raw.txt))
    - the same as your user

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **jankeromnes** after the PR has been reviewed.

Associated issue: *#4469*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

Associated issue: *#4469*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 1]
```diff
@@ -79,6 +108,14 @@ export class TeamDBImpl implements TeamDB {
             throw new Error('Please choose a team name containing only letters, numbers, -, _, \', or spaces.');
         }
         const slug = name.toLocaleLowerCase().replace(/[ ']/g, '-');
+        if (FORBIDDEN_SLUGS.indexOf(slug) !== -1) {
+            throw new Error('Creating a team with this name is not allowed');
```
Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:112]: **question**: This is probably out of the scope of these changes but could we skip the first part of the error message here as it does not seem much helpful for the user? What do you think? Happy to move this discussion to a new follow up issue. Maybe there're cases where the first part is helpful. 🤷 

| BEFORE | AFTER |
|-|-|
| Request createTeam failed with message: Creating a team with this name is not allowed | Creating a team with this name is not allowed |

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 2]
```diff
@@ -79,6 +108,14 @@ export class TeamDBImpl implements TeamDB {
...
+        }
+        const userRepo = await this.getUserRepo();
+        const existingUsers = await userRepo.query('SELECT COUNT(id) AS count FROM d_b_user WHERE fullName LIKE ? OR name LIKE ?', [ name, slug ]);
+        if (existingUsers[0].count > 0) {
+            throw new Error('A team cannot have the same name as an existing user');
```
Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:117]: **question**: Do we want to surface this kind of information? It seems this is privacy sensitive information, right? Maybe using a more generic message could be better? What do you think? 💭 

```suggestion
            throw new Error('Team path is already in use');
```

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 3]
```diff
@@ -15,6 +16,34 @@ import { DBTeamMembership } from "./entity/db-team-membership";
 import { DBUser } from "./entity/db-user";
 import { DBTeamMembershipInvite } from "./entity/db-team-membership-invite";
 
+const FORBIDDEN_SLUGS = [
```
Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:19]: **praise**: Thanks for extending the forbidden slugs here! 🌟

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 4]
```diff
@@ -15,6 +16,34 @@ import { DBTeamMembership } from "./entity/db-team-membership";
...
+    'setup',
+    'sorry',
+    'start',
+    'subscription',
+    'teams',
+    'upgrade-subscription',
+    'usage',
+    'variables',
+    'workspaces',
+    ...(blocklist),
```
Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:44]: **suggestion**: What do you think of adding a few more slugs that could beuseful in the near future?

For example, we could include `snapshots`, `prebuilds`, as well as `branches`, `issues`, `pull-requests`, and `merge-requests`. Your call! 🏓

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@
  * See License-AGPL.txt in the project root for license information.
  */
 
+import { list as blocklist } from "the-big-username-blacklist";
```
Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:7]: **praise**: Nice one! Thanks @corneliusludmann for also suggesting this in https://github.com/gitpod-io/gitpod/issues/4469#issuecomment-858657883. 🏀

**Review by gtsiolis — COMMENTED** [review]
Thanks for addign this @jankeromnes! 🔝 

Works as expected! ✔️ 

Left some minor comments below.

File: components/gitpod-db/src/typeorm/team-db-impl.ts (hunk 2 above)
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:117]: Hmm, not sure about the privacy aspect. If someone wants to create a team called "Jan Keromnes", and it fails, I'd be okay with the error message saying that a user is already called that name. 😋

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:117]: Also, user names come directly from GitHub or GitLab, where they're already public information anyway.

**Comment by jankeromnes** [comment]
Cool, thanks for the review! Will address your excellent suggestions. 🔝

/hold

File: components/gitpod-db/src/typeorm/team-db-impl.ts (hunk 2 above)
Comment (gtsiolis) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:117]: I still think this something we could consider changing in the future but it should be fine shipping this as is for now! 🙈

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:117]: What I'd love in the future is public Gitpod profiles, e.g. showcasing some public Gitpod stuff you'd like to share with the world. 🙂 Could give Gitpod more of a "social" vibe.

Comment (gtsiolis) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:117]: This is one of the reasons I'd vote for making this more privacy-aware but should be fine to skip this for now.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

Associated issue: *#4469*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by jankeromnes** [comment]
I believe all nits are addressed 😁

Screenshot 2021-08-10 at 15 51 55

@gtsiolis please take another look when you have time 👀

/unhold

**Comment by gtsiolis** [comment]
@jankeromnes UX works like a charm. 🔮 

1. Feel free to forward this to someone on the crew to take a deeper look in the code changes if needed before merging. 🏓 
2. We may need an approval from a dashboard owner as described in https://github.com/gitpod-io/gitpod/pull/5131#issuecomment-895988500 but not quite sure if this is the case after manual approval. 🤷

File: components/gitpod-db/src/typeorm/team-db-impl.ts (hunk 3 above)
Comment (csweichel) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:19]: How do we ensure this list does not drift w.r.t. the dashboard URLs?

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:19]: I don't think a perfect match is required at all times. The imported blocklist already contains a lot of teams names we wouldn't want to ever allow, and this complementary list ensures that current and potential near future dashboard paths aren't immediately taken by teams.

While a collision could still potentially happen in the more distant future, it's highly unlikely, and would be a sufficiently small problem that we can deal with it ad-hoc then.

**Comment by jankeromnes** [comment]
Many thanks for the reviews! 🙌 

> Works as expected! ✔️

> UX works like a charm. 🔮

Interpreting as /lgtm 😇

**Comment by gtsiolis** [comment]
Thanks @jankeromnes for merging this and @csweichel for taking a look! 👀

**Pull request merged**